sos/upload/__init__.py

```diff
@@ -0,0 +1,281 @@
+    """
+    This class is designed to upload files to a distribution
+    defined location. These files can be either sos reports,
+    sos collections, or other kind of files like: vmcores,
+    application cores, logs, etc.
+
+    """
+
+    desc = """Upload a file to a user or policy defined remote location"""
+
+    arg_defaults = {
+        'upload_file': '',
+        'case_id': '',
+        'low_priority': False,
+        'plugopts': [],
+        'upload_url': None,
+        'upload_directory': None,
+        'upload_user': None,
+        'upload_pass': None,
+        'upload_method': 'auto',
+        'upload_no_ssl_verify': False,
+        'upload_protocol': 'auto',
+        'upload_s3_endpoint': None,
+        'upload_s3_region': None,
+        'upload_s3_bucket': None,
+        'upload_s3_access_key': None,
+        'upload_s3_secret_key': None,
+        'upload_s3_object_prefix': None,
+        'upload_profile': None
+    }
+
+    def __init__(self, parser=None, args=None, cmdline=None, in_place=False,
+                 hook_commons=None, archive=None):
+        if not in_place:
+            # we are running `sos upload` directly
+            super().__init__(parser, args, cmdline)
+            self.from_cmdline = True
+        else:
+            # we are being hooked by either SoSReport or SoSCollector, don't
+            # re-init everything as that will cause issues, but instead load
+            # the needed bits from the calling component
+            self.opts = hook_commons['options']
+            self.policy = hook_commons['policy']
+            self.manifest = hook_commons['manifest']
+            self.parser = parser
+            self.cmdline = cmdline
+            self.args = args
+            self._upload_file = archive
+
+            self.ui_log = logging.getLogger('sos_ui')
+            self.from_cmdline = False
+            self.archive = archive
+        # add manifest section for upload
+        self.manifest.components.add_section('upload')
+
+        self.profiles_data = {
+            'redhat': {
+                'mod': "sos.upload.profiles.redhat",
+                'class': "RHELUpload",
+                'description': "Red Hat"
+            },
+            'ubuntu': {
+                'mod': "sos.upload.profiles.ubuntu",
+                'class': "UbuntuUpload",
+                'description': "Ubuntu"
+            },
+            'generic': {
+                'mod': "sos.upload.profiles",
+                'class': "Upload",
+                'description': "Generic"
+
+            }
+        }
+
+    @classmethod
+    def add_parser_options(cls, parser):
+        parser.usage = 'sos upload FILE [options]'
+        upload_grp = parser.add_argument_group(
+            'Upload Options',
+            'These options control how upload manages files'
+            )
+        upload_grp.add_argument("upload_file", metavar="FILE",
+                                help="The file or archive to upload")
```

Comment:
Would this PR intend to have the upload arguments inherited in other argument parsers, reducing code duplication?

Or does this require maintaining all three locations for consistency?
* `report/__init__.py`
* `collector/__init__.py`
* `upload/__init__.py`

Reply:
I hit this "code duplication problem" with arguments else-where, in collector/report/cleaner combination. We have similar duplication there that I am not happy with. Maybe worth to have some "component(s)<->parameters" abstraction / declaration, where say yaml or json file describes the allowed parameters and inheritance (of arguments and also components relations/inheritancy)?

But that is bit independent task to this PR.